File: 04/python/main.py
def score_game(boards, draws):
    def fill_if_contains(draw, board):
        for i, row in enumerate(board):
            for j, x in enumerate(row):
                if x == draw:
                    board[i][j] = None

    def is_winner(board):
        def columns(board):
            return ((row[i] for row in board) for i in range(len(board)))

        def any_seq_of_all_none(seqs):
            return any(all(x is None for x in seq) for seq in seqs)

        return any_seq_of_all_none(board) or any_seq_of_all_none(columns(board))

    def partition(seq, predicate):
        yes, no = [], []
        for x in seq:
            if predicate(x):
                yes.append(x)
            else:
                no.append(x)
        return yes, no

    def sum_remaining(board):
        return sum(x for row in board for x in row if x is not None)

    candidates = boards
    scores = []
    candidates = boards
    for draw in draws:
        for board in boards:
            fill_if_contains(draw, board)
        winners, candidates = partition(candidates, is_winner)
        for board in winners:
            scores.append(draw * sum_remaining(board))
    return scores


with open("../input.txt") as f:

    def parse_boards(lines):
        def parse_board(from_idx):
            def parse_row(row):
                return [int(s) for s in row.split() if s != ""]

            return [parse_row(lines[from_idx + i]) for i in range(5)]

        return [parse_board(i) for i in range(1, len(lines), 6)]

    lines = (line for line in f)
    draws = [int(s) for s in next(lines).strip().split(",")]
    remaining_lines = list(lines)
    boards = parse_boards(remaining_lines)
    scores = score_game(boards, draws)
    print("part 1:", scores[0])
    print("part 2:", scores[-1])

File: 04/rust/src/main.rs
use std::{num::ParseIntError, str::FromStr};

#[derive(Debug)]
struct BingoBoard([[Option<u8>; 5]; 5]);

impl FromStr for BingoBoard {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut result = [[Some(0u8); 5]; 5];
        for (i, line) in s.lines().enumerate() {
            for (j, num) in line.split_whitespace().enumerate() {
                result[i][j] = Some(num.parse::<u8>()?);
            }
        }
        Ok(BingoBoard(result))
    }
}

impl BingoBoard {
    fn mark(&mut self, num: u8) {
        for row in self.0.iter_mut() {
            for square in row.iter_mut() {
                if square == &Some(num) {
                    square.take();
                }
            }
        }
    }

    fn has_bingo(&self) -> bool {
        let grid = self.0;
        let horiontal_bingo = grid.iter().any(|row| row.iter().all(Option::is_none));
        let vertical_bingo = (0..5).any(|col| (0..5).all(|row| grid[row][col].is_none()));
        horiontal_bingo || vertical_bingo
    }

    fn get_unmarked_sum(&self) -> i32 {
        self.0
            .iter()
            .flat_map(|row| row.iter())
            .flat_map(Option::iter)
            .map(|&i| i as i32)
            .sum()
    }
}

fn score_game(mut boards: Vec<BingoBoard>, draws: &[u8]) -> Vec<i32> {
    let mut scores = Vec::new();
    for &draw in draws {
        for board in boards.iter_mut() {
            board.mark(draw);
        }
        for i in (0..boards.len()).rev() {
            if boards[i].has_bingo() {
                scores.push(draw as i32 * boards[i].get_unmarked_sum());
                boards.swap_remove(i);
            }
        }
    }
    scores
}

fn main() -> anyhow::Result<()> {
    let input = include_str!("../../input.txt");
    let mut input = input.split("\n\n");
    let draws = input.next().expect("Invalid input");

    let draws = draws
        .split(',')
        .map(u8::from_str)
        .collect::<Result<Vec<u8>, _>>()?;

    let boards = input.map(BingoBoard::from_str).collect::<Result<_, _>>()?;

    let scores = score_game(boards, &draws);
    println!("part 1: {}", scores[0]);
    println!("part 2: {}", scores[scores.len() - 1]);
    Ok(())
}
